File: server/sales_data2.py
from flask import Blueprint, jsonify, request
import pyodbc

sales_data2_bp = Blueprint('sales_data2', __name__)

connection_string = (
    'DRIVER={ODBC Driver 17 for SQL Server};'
    'SERVER=wsn-mis-068;'
    'DATABASE=codal;'
    'UID=sa;'
    'PWD=dbco@2023hamkaran'
)

@sales_data2_bp.route('/api/SalesData2', methods=['GET'])
def get_sales_data():
    company_name = request.args.get('companyName')
    query = "SELECT CompanyName, CompanyID, ReportDate, Value1, Value2, Value3 FROM mahane"
    params = ()

    if company_name:
        query += " WHERE CompanyName = ?"
        params = (company_name,)

    data = []
    with pyodbc.connect(connection_string) as conn:
        cursor = conn.cursor()
        cursor.execute(query, params)
        for row in cursor.fetchall():
            value1, value2, value3 = float(row.Value1), float(row.Value2), float(row.Value3)
            if value1 == 0:
                continue            
        
            data.append({
                "companyName": row.CompanyName,
                "companyID": row.CompanyID,
                "reportDate": row.ReportDate,
                "value1": value1/1000000,
                "value2": value2/1000000,
                "value3": value3/1000000,
                "percentage": value3/1000,
                "wow": 1 if (value1 > 0 and (value2 < 0 or value3 < 0)) else (-1 if (value1 < 0 and (value2 > 0 or value3 > 0)) else 0)
            })

    return jsonify(data)

@sales_data2_bp.route('/api/GetUrl2', methods=['POST'])
def get_url():
    data = request.get_json()
    company_name = data.get("companyName")

    query = "SELECT TOP 1 Url FROM mahane WHERE CompanyName like ?"

    with pyodbc.connect(connection_string) as conn:
        cursor = conn.cursor()
        cursor.execute(query, (company_name,))
        row = cursor.fetchone()
        if row:
            return jsonify({"url": row.Url})
        else:
            return jsonify({"url": ""})

File: go-codal-app/handlers/sales_data2.go
package handlers

import (
	"database/sql"
	"net/http"
	"strings"

	"go-codal-app/config"
	"go-codal-app/models"

	"github.com/gin-gonic/gin"
)

func GetSalesData2(c *gin.Context) {
	db := config.GetDB()
	defer db.Close()

	companyName := c.Query("companyName")
	query := "SELECT CompanyName, CompanyID, ReportDate, Value1, Value2, Value3 FROM mahane"

	var rows *sql.Rows
	var err error

	if companyName != "" {
		query += " WHERE CompanyName = @companyName"
		rows, err = db.Query(query, sql.Named("companyName", companyName))
	} else {
		rows, err = db.Query(query)
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer rows.Close()

	var data []models.SalesData2
	for rows.Next() {
		var s models.SalesData2
		var v1, v2, v3 sql.NullFloat64

		if err := rows.Scan(&s.CompanyName, &s.CompanyID, &s.ReportDate, &v1, &v2, &v3); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		s.Value1 = nullToFloat(v1)
		s.Value2 = nullToFloat(v2)
		s.Value3 = nullToFloat(v3)

		if s.Value1 == 0 {
			continue
		}

		s.Value1 /= 1_000_000
		s.Value2 /= 1_000_000
		s.Value3 /= 1_000_000
		s.Percentage = s.Value3 * 1000 // equivalent of: value3 / 1000

		if s.Value1 > 0 && (s.Value2 < 0 || s.Value3 < 0) {
			s.WoW = 1
		} else if s.Value1 < 0 && (s.Value2 > 0 || s.Value3 > 0) {
			s.WoW = -1
		} else {
			s.WoW = 0
		}

		data = append(data, s)
	}

	c.JSON(http.StatusOK, data)
}

func GetURL2(c *gin.Context) {
	db := config.GetDB()
	defer db.Close()

	var req models.GetURLRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	query := "SELECT TOP 1 Url FROM mahane WHERE CompanyName LIKE @companyName"
	likePattern := "%" + strings.TrimSpace(req.CompanyName) + "%"
	row := db.QueryRow(query, sql.Named("companyName", likePattern))

	var url sql.NullString
	if err := row.Scan(&url); err != nil {
		if err == sql.ErrNoRows {
			c.JSON(http.StatusOK, models.GetURLResponse{URL: ""})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if url.Valid {
		c.JSON(http.StatusOK, models.GetURLResponse{URL: url.String})
	} else {
		c.JSON(http.StatusOK, models.GetURLResponse{URL: ""})
	}
}
